File: ObjcPodDemo/RixEngineObjcDemo/UI/Admob/AdmobInterstitialVC.m
#import "AdmobInterstitialVC.h"
#import <GoogleMobileAds/GoogleMobileAds.h>
#import "AdConfig.h"

@interface AdmobInterstitialVC () <GADFullScreenContentDelegate>

@property (nonatomic, strong) UILabel *label;
@property (nonatomic, assign) BOOL isLoading;
@property (nonatomic, strong, nullable) GADInterstitialAd *interstitialAd;
@property (nonatomic, copy) NSString *TAG;

@end

@implementation AdmobInterstitialVC

- (void)viewDidLoad {
    [super viewDidLoad];
    self.TAG = @"Admob-interstitial:";
    self.view.backgroundColor = [UIColor whiteColor];
    self.navigationItem.title = NSLocalizedString(@"admob_interstitial", @"");

    UIButton *bnLoad = [self createButtonWithTitle:NSLocalizedString(@"load_ad", @"") action:@selector(buttonLoad)];
    [self.view addSubview:bnLoad];

    UIButton *bnShow = [self createButtonWithTitle:NSLocalizedString(@"show_ad", @"") action:@selector(buttonShow)];
    [self.view addSubview:bnShow];

    self.label = [self createLabel];
    [self.view addSubview:self.label];

    [NSLayoutConstraint activateConstraints:@[
        [bnLoad.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [bnLoad.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [bnLoad.heightAnchor constraintEqualToConstant:50],
        [bnLoad.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],

        [bnShow.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [bnShow.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [bnShow.heightAnchor constraintEqualToConstant:50],
        [bnShow.topAnchor constraintEqualToAnchor:bnLoad.bottomAnchor constant:20],

        [self.label.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [self.label.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [self.label.heightAnchor constraintEqualToConstant:50],
        [self.label.topAnchor constraintEqualToAnchor:bnShow.bottomAnchor constant:20],
    ]];
}

- (void)buttonLoad {
    NSLog(@"bnLoad");
    if (self.isLoading) {
        return;
    }
    [self updateUI:YES message:NSLocalizedString(@"loading", @"")];
    [self loadAd];
}

- (void)buttonShow {
    if (self.interstitialAd) {
        [self.interstitialAd presentFromRootViewController:self];
    } else {
        NSLog(@"Ad wasn't ready");
    }
}

- (void)loadAd {
    GADRequest *request = [GADRequest request];
    [GADInterstitialAd loadWithAdUnitID:[AdConfig Admob_Interstitial_Ad_Id]
                                request:request
                      completionHandler:^(GADInterstitialAd * _Nullable ad, NSError * _Nullable error) {
        if (error) {
            NSString *msg = [NSString stringWithFormat:@"%ld: %@", (long)error.code, error.localizedDescription];
            NSLog(@"%@ load: error: %@", self.TAG, msg);
            [self updateUI:NO message:[NSString stringWithFormat:NSLocalizedString(@"load_failed", @""), msg]];
            return;
        }
        NSLog(@"%@ load: success", self.TAG);
        [self updateUI:NO message:NSLocalizedString(@"load_success", @"")];
        self.interstitialAd = ad;
        self.interstitialAd.fullScreenContentDelegate = self;
    }];
}

- (void)updateUI:(BOOL)loading message:(NSString *)msg {
    self.isLoading = loading;
    self.label.text = msg;
}

#pragma mark - GADFullScreenContentDelegate

- (void)ad:(id<GADFullScreenPresentingAd>)ad didFailToPresentFullScreenContentWithError:(NSError *)error {
    NSString *msg = [NSString stringWithFormat:@"%ld: %@", (long)error.code, error.localizedDescription];
    NSLog(@"%@ ad: error:%@", self.TAG, msg);
}

- (void)adWillPresentFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
    NSLog(@"%@ adWillPresentFullScreenContent", self.TAG);
}

- (void)adDidDismissFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
    NSLog(@"%@ adDidDismissFullScreenContent", self.TAG);
}

- (void)adDidRecordImpression:(id<GADFullScreenPresentingAd>)ad {
    NSLog(@"%@ adDidRecordImpression", self.TAG);
}

- (void)adDidRecordClick:(id<GADFullScreenPresentingAd>)ad {
    NSLog(@"%@ adDidRecordClick", self.TAG);
}

- (void)adWillDismissFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
    NSLog(@"%@ adWillDismissFullScreenContent", self.TAG);
}

@end

File: SwiftPodDemo/AlxAdsDemo/UI/Admob/AdmobInterstitialVC.swift
//
//  AdmobInterstitialVC.swift
//  AdsDemo
//
//  Created by liu weile on 2023/5/30.
//

import UIKit
import GoogleMobileAds

class AdmobInterstitialVC: BaseUIViewController {
    
    private let TAG = "Admob-interstitial:"

    private var label:UILabel!

    private var isLoading:Bool=false

    private var interstitialAd:InterstitialAd?

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = .white
        navigationItem.title=NSLocalizedString("admob_interstitial", comment: "")

        let bnLoad=createButton(title: NSLocalizedString("load_ad", comment: ""),  action: #selector(buttonLoad))
        view.addSubview(bnLoad)

        let bnShow=createButton(title: NSLocalizedString("show_ad", comment: ""),  action: #selector(buttonShow))
        view.addSubview(bnShow)

        label=createLabel()
        view.addSubview(label)

        NSLayoutConstraint.activate([
            bnLoad.leftAnchor.constraint(equalTo: view.leftAnchor),
            bnLoad.rightAnchor.constraint(equalTo: view.rightAnchor),
            bnLoad.widthAnchor.constraint(equalTo: view.widthAnchor),
            bnLoad.heightAnchor.constraint(equalToConstant: 50),
            bnLoad.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),

            bnShow.leftAnchor.constraint(equalTo: view.leftAnchor),
            bnShow.rightAnchor.constraint(equalTo: view.rightAnchor),
            bnShow.widthAnchor.constraint(equalTo: view.widthAnchor),
            bnShow.heightAnchor.constraint(equalToConstant: 50),
            bnShow.topAnchor.constraint(equalTo: bnLoad.bottomAnchor, constant: 20),

            label.leftAnchor.constraint(equalTo: view.leftAnchor),
            label.rightAnchor.constraint(equalTo: view.rightAnchor),
            label.widthAnchor.constraint(equalTo: view.widthAnchor),
            label.heightAnchor.constraint(equalToConstant: 50),
            label.topAnchor.constraint(equalTo: bnShow.bottomAnchor, constant: 20),
           
        ])
    }
    
    @objc func buttonLoad(){
        print("bnLoad")
        if isLoading {
            return
        }

        updateUI(true, NSLocalizedString("loading", comment: ""))

        loadAd()
    }

    @objc func buttonShow(){
        if let ad=interstitialAd {
            ad.present(from: self)
        }else{
            print("Ad wasn't ready")
        }
    }
    
    func loadAd() {
        let request = Request()
        InterstitialAd.load(with:AdConfig.Admob_Interstitial_Ad_Id,
                           request: request,
                           completionHandler: { [self] ad, error in
            if let error = error {
                let error1 = error as NSError
                let msg="\(error1.code): \(error1.localizedDescription)"
                print("\(TAG) load: errror: \(msg)")
                updateUI(false, String(format: NSLocalizedString("load_failed", comment: ""), msg))
                return
            }
            print("\(TAG) load: success")
            updateUI(false, NSLocalizedString("load_success", comment: ""))
            self.interstitialAd = ad
            self.interstitialAd?.fullScreenContentDelegate=self
        })
    }
    
    func updateUI(_ loading:Bool,_ msg:String){
        self.isLoading=loading
        self.label.text=msg
    }

}

extension AdmobInterstitialVC:FullScreenContentDelegate{
    /// Tells the delegate that the ad failed to present full screen content.
    func ad(_ ad: FullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        let error1 = error as NSError
        let msg = "\(error1.code): \(error1.localizedDescription)"
        print("\(TAG) ad: error:\(msg)")
    }

    /// Tells the delegate that the ad will present full screen content.
    func adWillPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
        print("\(TAG) adWillPresentFullScreenContent")
    }

    /// Tells the delegate that the ad dismissed full screen content.
    func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        print("\(TAG) adDidDismissFullScreenContent")
    }
    
    func adDidRecordImpression(_ ad: any FullScreenPresentingAd) {
        print("\(TAG) adDidRecordImpression")
    }
    
    func adDidRecordClick(_ ad: any FullScreenPresentingAd) {
        print("\(TAG) adDidRecordClick")
    }
    
    func adWillDismissFullScreenContent(_ ad: any FullScreenPresentingAd) {
        print("\(TAG) adWillDismissFullScreenContent")
    }
    
}

